src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -485,27 +528,49 @@ private void ConsumeValue(byte marker)
                 JsonThrowHelper.ThrowJsonReaderException(ref this);
 #endif
             }
+            return true;
         }
 
-        private void ConsumeNumber()
+        private bool ConsumeNumber()
         {
             ValueType = JsonValueType.Number;
-            Value = GetNumber(_buffer.Slice(CurrentIndex));
+            if (!TryGetNumber(_buffer.Slice(CurrentIndex), out ReadOnlySpan<byte> number))
+                return false;
+            Value = number;
             CurrentIndex += Value.Length;
 #if UseInstrumented
             _position += Value.Length;
 #endif
+            return true;
         }
 
-        private void ConsumeNull()
+        private bool ConsumeNull()
         {
             Value = JsonConstants.NullValue;
             ValueType = JsonValueType.Null;
 
-            if (!_buffer.Slice(CurrentIndex).StartsWith(Value))
+            ReadOnlySpan<byte> span = _buffer.Slice(CurrentIndex);
+
+            Debug.Assert(span.Length > 0 && span[0] == Value[0]);
+
+            if (!span.StartsWith(Value))
             {
+                if (_isFinalBlock)
+                {
+                    goto Throw;
+                }
+                else
+                {
+                    if (span.Length > 1 && span[1] != Value[1])
+                        goto Throw;
```

Comment:
Should the reader be designed just like our encoders, i.e. they don't throw when they encounter bad data, but rather return with some "bad data" status. This allows the caller to inspect the bytesConsumed value to see where the bad data starts, fixup, and possibly keep going. This is an honest question, i.e. I am not sure what we should do here. @jamesNk, do people every try to recover from bad JSON when parsing?

cc: @davidfowl, @GrabYourPitchforks

Reply:
No I think it should just throw.

Reply:
There is a feature in Json.NET where you can attempt to catch an exception and retry. I found it caused more problems that it was worth.

I'd throw.

Reply:
> Should the reader be designed just like our encoders, i.e. they don't throw when they encounter bad data, but rather return with some "bad data" status. This allows the caller to inspect the bytesConsumed value to see where the bad data starts, fixup, and possibly keep going. 

It would definitely provide more flexibility to the user, at the cost of additional complexity. Since we don't have a destination here, we have three states: Done, Invalid, NeedsMore. If throwing for invalid is the expected behaviour, having an OperationStatus instead of a bool for the two states seems unnecessary.